### PR TITLE
Documentation for typed EventStream 

### LINKED ISSUE #1162: EventStream documentation on Typed is required
Pekko has EventBus(low-level) and EventStream documentation on Classic: https://pekko.apache.org/docs/pekko/current/event-bus.html

But there is no documentation on typed at present.

### PR DESCRIPTION
**Pull request by sfali**
For #1162

### CODE REVIEW

File: actor-typed-tests/src/test/java/org/apache/pekko/actor/typed/eventstream/EventStreamDocTest.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
+package org.apache.pekko.actor.typed.eventstream;
```
Comment (pjfanning) [inline actor-typed-tests/src/test/java/org/apache/pekko/actor/typed/eventstream/EventStreamDocTest.java:1]: can you add a standard Apache header here? not the one that mentions Akka

File: actor-typed-tests/src/test/java/org/apache/pekko/actor/typed/eventstream/EventStreamSuperClassDocTest.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
+package org.apache.pekko.actor.typed.eventstream;
```
Comment (pjfanning) [inline actor-typed-tests/src/test/java/org/apache/pekko/actor/typed/eventstream/EventStreamSuperClassDocTest.java:1]: likewise - all source files must have an Apache header

File: actor-typed-tests/src/test/java/org/apache/pekko/actor/typed/eventstream/EventStreamDocTest.java (hunk 1 above)
Comment (sfali) [reply actor-typed-tests/src/test/java/org/apache/pekko/actor/typed/eventstream/EventStreamDocTest.java:18]: Added

**Review by pjfanning — APPROVED** [review]
lgtm

File: actor-typed-tests/src/test/java/org/apache/pekko/actor/typed/eventstream/EventStreamSuperClassDocTest.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+        public AllDeadLettersListenerBehavior(ActorContext<Command> context) {
+            super(context);
+        }
+
+        @Override
+        public Receive<Command> createReceive() {
+            return newReceiveBuilder().onMessage(AllDeadLettersWrapper.class, msg -> {
+                final AllDeadLetters allDeadLetters = msg.allDeadLetters;
+                final Class<? extends AllDeadLetters> klass = msg.allDeadLetters.getClass();
+                if (klass.isAssignableFrom(DeadLetter.class)) {
```
Comment (Roiocam) [inline actor-typed-tests/src/test/java/org/apache/pekko/actor/typed/eventstream/EventStreamSuperClassDocTest.java:79]: The actor already has done similar works on `Receive`, why do we need those type-checks on here?

I don't think this programming style is something we want to share with users.

In my opinion, the EventStream use case mostly was embedded Pub/Sub feature(aka EventBus), we treat the type as a Topic(e.g. `AllDeadLetters`), and the instance with subtype was the specific message(e.g. `DeadLetter`, `Dropped`, `UnhandledMessage`).

Or a smaller range of channels: `DeadLetter` is used as both a topic and message instance type.

In your PR, subscribing to Command doesn't make sense. When users need to publish, they should convert AllDeadLetter to AllDeadLetterWrapper. This is not the correct usage.

In general, if you want this Actor to receive messages from other Actors (such as in your case), you can use Adapter, but it is actually like this:


```scala
 ActorSystem<Command> system = ActorSystem.create(AllDeadLettersListenerBehavior.create(), "AllDeadLettersListener");
// get adapter ref, or do subscribe logic in actor internal.
val adapterRef = system ? getAllDeadLettersAdapter
// subscribe the raw type of `AllDeadLetters`, but receive with adapterRef
system.eventStream().tell(new EventStream.Subscribe<>(AllDeadLetters.class, adapterRef));
// and then publish look like this.
system.eventStream().tell(new EventStream.Publish<>(DeadLetter("msg", senderRef, recipientRef)));

// In your code, publish in this way, but in fact, you can't expect the producer to use the type of message that the consumer can receive. On the contrary, the consumer must adapt to the type of producer, which is very important for the ability to broadcast.
val deadLetter = DeadLetter("msg", senderRef, recipientRef)
system.eventStream().tell(new EventStream.Publish<>(new AllDeadLettersWrapper(deadLetter)));
```

**Review by pjfanning — DISMISSED** [review]
Build is broken

File: actor-typed-tests/src/test/java/org/apache/pekko/actor/typed/eventstream/EventStreamSuperClassDocTest.java (hunk 3 above)
Comment (sfali) [reply actor-typed-tests/src/test/java/org/apache/pekko/actor/typed/eventstream/EventStreamSuperClassDocTest.java:83]: Hmm, let me take a look.

Comment (sfali) [reply actor-typed-tests/src/test/java/org/apache/pekko/actor/typed/eventstream/EventStreamSuperClassDocTest.java:90]: I had the missing subscription in the constructor of `AllDeadLettersListenerBehavior` and there is no need to have subscription in the client side, which is removed now.

**Comment by sfali** [comment]
> Build is broken

Not sure what is broken, I am able to compile locally.

**Comment by mdedetrich** [comment]
> > Build is broken
> 
> Not sure what is broken, I am able to compile locally.

The build for Pekko core is a bit complicated as we use different JDK's to produce different outputs. In this case its generating docs with JKD11 using JDK 9 specific classes that is causing the failure, specifically running `sbt ";+TestJdk9 / compile ; +compile:doc"` using JDK 11.

Doing so produces the following error

```
[error] /home/runner/work/incubator-pekko/incubator-pekko/actor-typed-tests/src/test/scala/org/apache/pekko/actor/typed/eventstream/EventStreamDocSpec.scala:45:27: overloaded method value info with alternatives:
[error]   (x$1: org.slf4j.Marker,x$2: String,x$3: Object*)Unit <and>
[error]   (x$1: org.slf4j.Marker,x$2: String,x$3: Any,x$4: Any)Unit <and>
[error]   (x$1: String,x$2: Object*)Unit
[error]  cannot be applied to (String, String, String, Any)
[error]               context.log.info("Dead letter received from sender ({}) to recipient ({}) with message: {}",
[error]                           ^
[error] /home/runner/work/incubator-pekko/incubator-pekko/actor-typed-tests/src/test/scala/org/apache/pekko/actor/typed/eventstream/EventStreamDocSpec.scala:69:31: overloaded method value info with alternatives:
[error]   (x$1: org.slf4j.Marker,x$2: String,x$3: Object*)Unit <and>
[error]   (x$1: org.slf4j.Marker,x$2: String,x$3: Any,x$4: Any)Unit <and>
[error]   (x$1: String,x$2: Object*)Unit
[error]  cannot be applied to (String, String, String, Any)
[error]                   context.log.info("DeadLetter received from sender ({}) to recipient ({}) with message: {}",
[error]                               ^
[error] /home/runner/work/incubator-pekko/incubator-pekko/actor-typed-tests/src/test/scala/org/apache/pekko/actor/typed/eventstream/EventStreamDocSpec.scala:73:31: overloaded method value info with alternatives:
[error]   (x$1: org.slf4j.Marker,x$2: String,x$3: Object*)Unit <and>
[error]   (x$1: String,x$2: Object*)Unit
[error]  cannot be applied to (String, String, String, String, Any)
[error]                   context.log.info("Dropped: sender ({}) to recipient ({}) with message: {}, reason: {}",
[error]                               ^
[error] /home/runner/work/incubator-pekko/incubator-pekko/actor-typed-tests/src/test/scala/org/apache/pekko/actor/typed/eventstream/EventStreamDocSpec.scala:82:31: overloaded method value info with alternatives:
[error]   (x$1: org.slf4j.Marker,x$2: String,x$3: Object*)Unit <and>
[error]   (x$1: org.slf4j.Marker,x$2: String,x$3: Any,x$4: Any)Unit <and>
[error]   (x$1: String,x$2: Object*)Unit
[error]  cannot be applied to (String, String, String, Any)
[error]                   context.log.info("UnhandledMessage received from sender ({}) to recipient ({}) with message: {}",
[error]                               ^
[error] four errors found
```

So its an issue with unresolved resolution of overloaded `log.info`

**Comment by mdedetrich** [comment]
@sfali Are you done with the commits so I can run the PR?

**Comment by mdedetrich** [comment]
There are still issues

```
[error] -- Error: /home/runner/work/incubator-pekko/incubator-pekko/actor-typed-tests/src/test/scala/org/apache/pekko/actor/typed/eventstream/EventStreamDocSpec.scala:40:59 
[error] 40 |          val adapter = context.messageAdapter[DeadLetter](DeadLetterWrapper)
[error]    |                                                           ^^^^^^^^^^^^^^^^^
[error]    |The method `apply` is inserted. The auto insertion will be deprecated, please write `org.apache.pekko.actor.typed.eventstream.EventStreamDocSpec.DeadLetterListener.
[error]    |  DeadLetterWrapper.apply` explicitly.
[error] -- Error: /home/runner/work/incubator-pekko/incubator-pekko/actor-typed-tests/src/test/scala/org/apache/pekko/actor/typed/eventstream/EventStreamDocSpec.scala:62:63 
[error] 62 |          val adapter = context.messageAdapter[AllDeadLetters](AllDeadLettersWrapper)
[error]    |                                                               ^^^^^^^^^^^^^^^^^^^^^
[error]    |The method `apply` is inserted. The auto insertion will be deprecated, please write `org.apache.pekko.actor.typed.eventstream.EventStreamDocSpec.AllDeadLetterListener
[error]    |  .AllDeadLettersWrapper.apply` explicitly.
```

Just need to `.apply`

**Comment by sfali** [comment]
> There are still issues
> 
> ```
> [error] -- Error: /home/runner/work/incubator-pekko/incubator-pekko/actor-typed-tests/src/test/scala/org/apache/pekko/actor/typed/eventstream/EventStreamDocSpec.scala:40:59 
> [error] 40 |          val adapter = context.messageAdapter[DeadLetter](DeadLetterWrapper)
> [error]    |                                                           ^^^^^^^^^^^^^^^^^
> [error]    |The method `apply` is inserted. The auto insertion will be deprecated, please write `org.apache.pekko.actor.typed.eventstream.EventStreamDocSpec.DeadLetterListener.
> [error]    |  DeadLetterWrapper.apply` explicitly.
> [error] -- Error: /home/runner/work/incubator-pekko/incubator-pekko/actor-typed-tests/src/test/scala/org/apache/pekko/actor/typed/eventstream/EventStreamDocSpec.scala:62:63 
> [error] 62 |          val adapter = context.messageAdapter[AllDeadLetters](AllDeadLettersWrapper)
> [error]    |                                                               ^^^^^^^^^^^^^^^^^^^^^
> [error]    |The method `apply` is inserted. The auto insertion will be deprecated, please write `org.apache.pekko.actor.typed.eventstream.EventStreamDocSpec.AllDeadLetterListener
> [error]    |  .AllDeadLettersWrapper.apply` explicitly.
> ```
> 
> Just need to `.apply`

Updated

**Comment by sfali** [comment]
Finally passed

**Comment by sfali** [comment]
> @sfali Are you done with the commits so I can run the PR's CI?

Yes, I am done

**Review by mdedetrich — APPROVED** [review]
lgtm, @pjfanning can you recheck

**Review by pjfanning — APPROVED** [review]
lgtm

**Comment by pjfanning** [comment]
merged - thanks @sfali

**Pull request merged**